lighthouse-core/runner.js

```diff
@@ -19,169 +18,168 @@ const path = require('path');
 const URL = require('./lib/url-shim');
 const Sentry = require('./lib/sentry');
 
-class Runner {
-  static run(connection, opts) {
-    // Clean opts input.
-    opts.settings = opts.config && opts.config.settings || {};
-
-    // List of top-level warnings for this Lighthouse run.
-    const lighthouseRunWarnings = [];
+const Connection = require('./gather/connections/connection.js'); // eslint-disable-line no-unused-vars
 
-    // save the initialUrl provided by the user
-    opts.initialUrl = opts.url;
-    if (typeof opts.initialUrl !== 'string' || opts.initialUrl.length === 0) {
-      return Promise.reject(new Error('You must provide a url to the runner'));
-    }
+/** @typedef {LH.Result & {artifacts: LH.Artifacts}} RunnerResult */
 
-    let parsedURL;
+class Runner {
+  /**
+   * @param {Connection} connection
+   * @param {{config: LH.Config, url: string, initialUrl?: string, driverMock?: Driver}} opts
+   * @return {Promise<RunnerResult|undefined>}
+   */
+  static async run(connection, opts) {
```

Comment:
`opts` now never leaves `run() and `_gatherArtifactsFromBrowser()`, so we have a lot more freedom to do things like refactor our `url`/`initialUrl`/`finalUrl` business (not to mention `options.settings` (now isolated to `_gatherArtifactsFromBrowser()`) vs `options.config.settings`, etc)

Reply:
very impressed with your commitment to appropriately back-ticking tokens in this sentence. 
👨‍💻 ⌨️ 🥇